src/irmin/store.ml

```diff
@@ -1173,9 +1173,9 @@ module Make (B : Backend.S) = struct
                         Heap.add heap (commit, current_depth + 1)
                     in
                     let tree = Commit.tree commit in
-                    let+ e = Tree.find tree key in
-                    match (e, current_value) with
-                    | Some x, Some y -> not (equal_contents x y)
+                    let+ e = Tree.find_tree tree key in
+                    match (e, current_tree) with
+                    | Some x, Some y -> Tree.hash x <> Tree.hash y
```

Comment:
Using the generic `<>` function isn't guaranteed to work for some hash representations, maybe consider using an `equal_hash` function instead? (as done in other modules)